### PR TITLE
Add Future Lens script

### PR DESCRIPTION
**Pull request by tvhong**
### Notes
* Fix a few bugs in the soft prompt implementation.
* Allow 2d or 3d tensors as soft prompt.
* Add `scripts/future_lens.py`.

### Testing
I successfully reproduce the result from `future_lens` paper with `remote=True`:
```
(obvs) Parrots-MacBook-Pro :: workplace/obvslib/scripts ‹vhong/fix_soft_prompt*› % python future_lens.py
660cf1f37d1a28a42c970913 - RECEIVED: Your job has been received and is waiting approval.
660cf1f37d1a28a42c970913 - APPROVED: Your job was approved and is waiting to be run.
660cf1f37d1a28a42c970913 - COMPLETED: Your job has been completed.
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
Downloading result: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.03M/1.03M [00:00<00:00, 2.06MB/s]
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.
A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.
660cf1f7bd925a617fa3fd61 - RECEIVED: Your job has been received and is waiting approval.
660cf1f7bd925a617fa3fd61 - APPROVED: Your job was approved and is waiting to be run.
660cf1f7bd925a617fa3fd61 - COMPLETED: Your job has been completed.
Downloading result: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.62M/2.62M [00:00<00:00, 4.07MB/s]
['_', ' _', ' _', ' _', ' _', ' _', ' _', ' _', ' _', ' _', ' Back', ' to', ' the', ' Future']

```

### CODE REVIEW

File: obvs/patchscope.py [hunk 1]
```diff
@@ -181,14 +188,16 @@ def source_forward_pass(self) -> None:
...
-        with self.source_model.trace(self.source.text_prompt, remote=self.REMOTE) as _:
+        with self.source_model.trace(self.source.text_prompt, remote=self.REMOTE):
             if self.source.soft_prompt is not None:
                 # TODO: validate this with non GPT2 & GPTJ models
                 self.source_model.transformer.wte.output = self.source.soft_prompt
 
             self._source_hidden_state = self.manipulate_source().save()
             self.source_output = self.source_model.lm_head.output[0].save()
 
+        self._source_hidden_state = self._source_hidden_state.value
```
Comment (llinauer) [inline obvs/patchscope.py:199]: Out of curiosity; why do we need this line?

File: scripts/future_lens.py
Comment (llinauer) [inline scripts/future_lens.py:1]: Loading the soft prompt fails if we are not in the `scripts` directory.

Could we maybe make the path independent by using something like:
cwd = os.path.dirname(os.path.abspath(__file__)) 
-> torch.load(cwd)

File: obvs/patchscope.py (hunk 1 above)
Comment (tvhong) [reply obvs/patchscope.py:199]: I need it when running nnsight with remote=True.

Without extracting the value tensor here, the target run will send the proxy object to the remote server, and nnsight doesn’t send the correct underlying value.

Maybe we should create an issue in nnsight 🤔

File: scripts/future_lens.py
Comment (tvhong) [reply scripts/future_lens.py:1]: Good point. I’ll make a commit.

Comment (tvhong) [reply scripts/future_lens.py:1]: Done.

Tested by running `python future_lens.py` and `python scripts/future_lens.py`.

File: obvs/patchscope.py (hunk 1 above)
Comment (tvhong) [reply obvs/patchscope.py:199]: Here is the nnsight issue: https://github.com/ndif-team/nnsight/issues/104

Comment (llinauer) [reply obvs/patchscope.py:199]: I see. No further questions, let's merge.

**Pull request merged**